lib/cdo/unicorn_listener.rb

```diff
@@ -0,0 +1,105 @@
+  # `queued` - the number of queued TCP/socket requests
+  # `calling` - the maximum number of currently-executing requests at any point during the interval.
+  #
+  # The :listeners option accepts an array of strings (TCP or Unix domain socket names).
+  # By default, all listeners used by the Unicorn master process are monitored.
+  #
+  # Any errors are forwarded to Honeybadger for logging and notifying.
+  class UnicornListener < Raindrops::Middleware
+    def initialize(app, opts = {})
+      super(app, opts.merge(stats: StatsWithMax.new))
+      interval = opts[:interval] || 1
+      report_count = opts[:report_count] || 60
+
+      task = reporting_task(interval, report_count)
+      task.add_observer {|_, _, ex| Honeybadger.notify(ex) if ex}
+      task.execute
+    end
+
+    # Periodically collect unicorn-listener metrics every `delay` seconds,
+    # reporting every time `report_count` metrics have been collected.
+    # @param interval [Fixnum]
+    # @param report_count [Fixnum]
+    def reporting_task(interval, report_count)
+      stat_names = %i(active queued calling)
+      stats = stat_names.map {|name| [name, []]}.to_h
+      Concurrent::TimerTask.new(execution_interval: interval) do
+        stat_values = collect_listener_stats + [@stats.max_calling.tap {@stats.reset_max}]
+        stats.zip(stat_values) {|stat, val| stat[1] << {timestamp: Time.now, value: val}}
+        report!(stats) if stats.values.first.count >= report_count
+      end
+    end
+
+    # Collect current snapshot of tcp/unix listener stats.
+    def collect_listener_stats
+      stats = {}
+      stats.merge! Raindrops::Linux.tcp_listener_stats(@tcp.uniq) if @tcp
+      stats.merge! Raindrops::Linux.unix_listener_stats(@unix.uniq) if @unix
+      %i(active queued).map do |name|
+        stats.values.map(&name).inject(:+)
+      end
+    end
+
+    # Report all stats as CloudWatch metrics, then clear the collection.
+    # @param stats [Hash]
+    def report!(stats)
+      metric_data = stats.map do |name, stat|
+        stat.reject {|datum| datum[:value].nil?}.map do |datum|
+          {
+            metric_name: name,
+            dimensions: [
+              {name: "Environment", value: CDO.rack_env},
+              {name: "Host", value: CDO.pegasus_hostname}
+            ],
+            timestamp: datum[:timestamp],
+            value: datum[:value],
+            unit: 'Count',
+            storage_resolution: 1
+          }
+        end
+      end.flatten
+      stats.values.map(&:clear)
+      Cdo::Metrics.push('Unicorn', metric_data)
+    end
+  end
+
+  # Adds an atomic `max_calling` counter to Raindrops::Middleware::Stats.
+  # This allows us to record the peak number of currently-executing
+  # worker-processes at any point, which a one-second sampling interval
+  # might not otherwise capture.
+  class StatsWithMax < Raindrops::Middleware::Stats
+    def initialize
+      super
+      @raindrops.size = 3
```

Comment:
Make the 3 and 2's be a constant and not a magic #?
Also, I don't understand what's going on here... is `@raindrops` a base class thing where @raindrops[0] and [1] are already in use, and we are expanding one more slot for our use? Some more comments would be helpful.

Reply:
yeah, this is specific to the class ([`Raindrops::Middleware::Stats`](http://repo.or.cz/raindrops.git/blob/HEAD:/lib/raindrops/middleware.rb#l80)) this class is extending, which initializes `:calling` and `:writing` as `Raindrops` counters (which are backed by the `@raindrops` instance variable. Here I'm increasing the size by 1 to introduce a third variable. I can add more comments / make less use of magic integers in a future commit.

Reply:
I managed to simplify the implementation in a way that doesn't require interacting with this internal instance-variable or using any index constants.